File: solutions/041.First_Missing_Positive/AC_array_n.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_array_n.cpp
*  Create Date: 2015-01-28 09:38:55
*  Descripton:  Brute force. This will use O(n) space && O(n) time.
*/

#include <bits/stdc++.h>

using namespace std;
const int N = 0;

class Solution {
public:
    int firstMissingPositive(int A[], int n) {
        vector<bool> rec(n + 2);
        for (int i = 0; i < n; i++) {
            if (A[i] < n + 2 && A[i] > 0)
                rec[A[i]] = true;
        }
        for (int i = 1; i < n + 2; i++)
            if (!rec[i])
                return i;
    }
};

int main() {
    int n, A[100];
    Solution s;
    cin >> n;
    for (int i = 0; i < n; i++)
        cin >> A[i];
    cout << s.firstMissingPositive(A, n) << endl;
    return 0;
}

File: solutions/238.Product_of_Array_Except_Self/AC_array_n.java
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_array_n.java
*  Create Date: 2016-02-24 23:23:27
*  Descripton:   
*/

public static class Solution {
    public int[] productExceptSelf(int[] nums) {
        if (nums.length == 0) return nums;
        int[] left = new int[nums.length];
        left[0] = 1;
        for (int i = 1; i < nums.length; ++i) {
            left[i] = left[i - 1] * nums[i - 1];
        }
        int right = 1;
        for (int i = nums.length - 2; i >= 0; --i) {
            right *= nums[i + 1];
            left[i] *= right;
        }
        return left;
    }
}
